rpc/amqp-rpc.go

```diff
@@ -407,7 +411,31 @@ func (rpc *AmqpRPCServer) Start(c cmd.Config) error {
 			select {
 			case msg, ok := <-msgs:
 				if ok {
-					rpc.processMessage(msg)
+					if rpc.maxGoroutines > 0 && atomic.LoadInt64(&rpc.currentGoroutines) >= rpc.maxGoroutines {
+						response := RPCResponse{Error: wrapError(core.InternalServerError("RPC server has spawned too many Goroutines"))}
+						jsonResponse, err := json.Marshal(response)
+						if err != nil {
+							// AUDIT[ Error Conditions ] 9cc4d537-8534-4970-8665-4b382abe82f3
+							rpc.log.Audit(fmt.Sprintf(" [s>][%s][%s] Error condition marshalling RPC response %s [%s]", rpc.serverQueue, msg.ReplyTo, msg.Type, msg.CorrelationId))
+							break // This breaks the select case not the for loop
+						}
+						rpc.Channel.Publish(
+							AmqpExchange,
+							msg.ReplyTo,
+							AmqpMandatory,
+							AmqpImmediate,
+							amqp.Publishing{
+								CorrelationId: msg.CorrelationId,
+								Type:          msg.Type,
+								Body:          jsonResponse,
```

Comment:
We can probably put a lower expiration on these, like 1 second.

Also, break out this chunk into a special function, like replyTooManyRequests.

Not for this change, but: Instead of counting and then erroring, maybe we should stop pulling requests off the channel when we are overloaded. We should check that ceasing to read from the go channel will actually make our AMQP library stop pulling from the queue, but hopefully it will.